redis/redis.go

```diff
@@ -0,0 +1,187 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/0xERR0R/blocky/config"
+	"github.com/0xERR0R/blocky/log"
+	"github.com/0xERR0R/blocky/model"
+	"github.com/go-redis/redis/v8"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	CacheChannelName string = "blocky_cache_sync"
+	CacheStorePrefix string = "blocky:cache:"
+)
+
+type CacheMessage struct {
+	Key      string
+	Response *model.Response
+}
+
+// Client for redis communication
+type Client struct {
+	config  *config.RedisConfig
+	client  *redis.Client
+	l       *logrus.Entry
+	ctx     context.Context
+	Channel chan *CacheMessage
+}
+
+// New creates a new redis client
+func New(cfg *config.RedisConfig) (*Client, error) {
+	// disable redis if no address is provided
+	if cfg == nil || len(cfg.Address) == 0 {
+		return nil, nil
+	}
+
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     cfg.Address,
+		Password: cfg.Password,
+		DB:       cfg.Database,
+	})
+	ctx := context.Background()
+
+	var err error
+
+	attempt := 1
+	for attempt <= cfg.ConnectionAttempts {
+		_, err = rdb.Ping(ctx).Result()
+		if err == nil {
+			// construct client
+			res := &Client{
+				config:  cfg,
+				client:  rdb,
+				l:       log.PrefixedLog("redis"),
+				ctx:     ctx,
+				Channel: make(chan *CacheMessage),
+			}
+
+			// start listener
+			err = res.startSubscriptionListener()
+
+			return res, err
+		}
+
+		time.Sleep(time.Duration(cfg.ConnectionCooldown))
+		attempt++
+	}
+
+	return nil, err
+}
+
+// PublishCache publish cache to redis async
+func (c *Client) PublishCache(key string, response *model.Response) {
+	msg, errConv := response.ConvertToCache(key)
+	if errConv == nil {
+		binMsg, errMar := msg.MarshalBinary()
+		if errMar == nil {
+			c.client.Publish(c.ctx, CacheChannelName, binMsg)
+			c.client.Set(c.ctx, prefixKey(key), binMsg, time.Duration(0))
```

Comment:
Thought about it too but would solve it a little different:
- create a configuration option to let the user configure redis.cacheTTL
- set the dns result ttl to 0 upon redis insertion
- set the prefetch flag(if enabled config wise) upon cache insertion following a sync message

Otherwise we would create the same issue as in #290 in redis.

Maybe you got a different idea to combat this issue?

Reply:
hm, I don't understand your approach. Each DNS response has a TTL time. This TTL should be propagated via redis to other blocky instance otherwise there would be a cache entry in blocky with wrong TTL. We can try to use TTL in redis -> this would also cleanup expired cache entries in redis. Or we put "expirationTimestamp" into the cache struct, so the other blocky instance can calculate TTL by something like "expirationTimestamp - now())